File: java/src/eu/semaine/jms/message/SEMAINEXMLMessage.java
package eu.semaine.jms.message;

import java.io.StringReader;

import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.TextMessage;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

import org.w3c.dom.Document;
import org.xml.sax.InputSource;

import eu.semaine.exceptions.MessageFormatException;

public class SEMAINEXMLMessage extends SEMAINEMessage
{
	public static final String IS_XML = "xml";
	
	protected Document doc = null;
	
	public SEMAINEXMLMessage(Message message)
	throws MessageFormatException
	{
		super(message);
		if (!(message instanceof TextMessage)) {
			throw new MessageFormatException("Expected a text message, but got a "+message.getClass().getName());
		}
		try {
			// the following will return false if the property is not set
			// or if it does not contain the value "true".
			boolean isXML = message.getBooleanProperty(IS_XML);
			if (!isXML) throw new MessageFormatException("expected XML message");
		} catch (JMSException j) {
			throw new MessageFormatException("Cannot read property '"+IS_XML+"'", j);
		}
		parseDocument();
	}
	
	public Document getDocument()
	throws MessageFormatException
	{
		return doc;
	}
	
	protected void parseDocument()
	throws MessageFormatException
	{
		try {
			DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
			factory.setNamespaceAware(true);
			DocumentBuilder builder = factory.newDocumentBuilder();
			doc = builder.parse(new InputSource(new StringReader(getText())));
		} catch (Exception e) {
			throw new MessageFormatException("Cannot parse XML message", e);
		}
	}
	
}

File: c++/src/semaine/cms/message/SEMAINEXMLMessage.cpp
/*
 *  SEMAINEXMLMessage.cpp
 *  semaine
 *
 *  Created by Marc Schröder on 19.09.08.
 *  Copyright 2008 DFKI GmbH. All rights reserved.
 *
 */

#include "SEMAINEXMLMessage.h"

using namespace XERCES_CPP_NAMESPACE;

namespace semaine {
namespace cms {
namespace message {

const std::string SEMAINEXMLMessage::IS_XML = "xml";


SEMAINEXMLMessage::SEMAINEXMLMessage(const Message * message)
throw(semaine::cms::exceptions::MessageFormatException,semaine::cms::exceptions::SystemConfigurationException) :
	SEMAINEMessage(message), document(NULL)
{
	const TextMessage * tm = dynamic_cast<const TextMessage *>(message);
	if (tm == NULL) {
		throw semaine::cms::exceptions::MessageFormatException(std::string("Expected a text message, but got a ")+typeid(*message).name());
	}
	
	try {
		bool isXML = message->getBooleanProperty(IS_XML);
		if (!isXML)
			throw semaine::cms::exceptions::MessageFormatException("expected XML message");
	} catch (const CMSException & ce) {
		ce.printStackTrace();
		throw semaine::cms::exceptions::MessageFormatException(std::string("Cannot read property '")+IS_XML+"'");
	}
	parseDocument();
}

SEMAINEXMLMessage::~SEMAINEXMLMessage()
{
	if (document != NULL) {
		document->release();
		document = NULL;
	}
}

void SEMAINEXMLMessage::parseDocument()
throw(semaine::cms::exceptions::MessageFormatException,semaine::cms::exceptions::SystemConfigurationException)
{

    try {
		std::string msgText = getText();
	    document = XMLTool::parse(msgText);
    } catch (const XMLException& toCatch) {
        char* err = XMLString::transcode(toCatch.getMessage());
        std::cerr << err << std::endl;
        XMLString::release(&err);
        throw semaine::cms::exceptions::MessageFormatException("Cannot parse xml");
	} catch (const DOMException& toCatch) {
        char* err = XMLString::transcode(toCatch.msg);
        std::cerr << err << std::endl;
        XMLString::release(&err);
        throw semaine::cms::exceptions::MessageFormatException("Cannot parse xml");
    } catch (...) {
        throw semaine::cms::exceptions::MessageFormatException("Cannot parse xml");
    }


}


} // namespace message
} // namespace cms
} // namespace semaine
